lib/collection/src/wal.rs

```diff
@@ -110,8 +110,16 @@ impl<'s, R: DeserializeOwned + Serialize + Debug> SerdeWal<R> {
             .map_err(|err| WalError::WriteWalError(format!("{err:?}")))
     }
 
-    pub fn read_all(&'s self) -> impl Iterator<Item = (u64, R)> + 's {
-        self.read(self.first_index())
+    pub fn read(&self, start_index: u64) -> impl DoubleEndedIterator<Item = (u64, R)> + '_ {
+        (start_index..self.last_index()).map(move |idx| {
```

Comment:
I thought the `last_index()` is inclusive. Which is interesting, because if that is the case, it means the previous implementation is broken too.

```suggestion
        (start_index..=self.last_index()).map(move |idx| {
```

Note: I'm investigating!

Reply:
This is the case indeed! :exploding_head: 

```rust
    /// Last index that is still available in logical WAL.
    pub fn last_index(&self) -> u64 {
        self.wal.last_index()
    }
```

which points to:

```rust
    /// The index of the last entry
    pub fn last_index(&self) -> u64 {
        let num_entries = self.num_entries();
        self.first_index() + num_entries.saturating_sub(1)
    }
```

The `num_entries.saturating_sub(1)` means the higher bound is inclusive.

So we want to loop over that records as well:

```diff
-(start_index..self.last_index())
+(start_index..=self.last_index())
```

Reply:
Yeah, that's probably just a bug I've introduced. Thanks. (Maybe that's why tests failed. This PR is more of idea dump atm.)

Reply:
Fixed this, to see if it will fix CI.